File: src/mentoai/pipeline/silver.py
import asyncio
import logging
import re
from typing import Any

import polars as pl

from mentoai.db.pool import executemany, fetch

logger = logging.getLogger(__name__)

COLUMNS = [
    "source",
    "source_id",
    "company",
    "position",
    "location",
    "intro",
    "main_tasks",
    "requirements",
    "preferred_points",
    "benefits",
    "employment_type",
    "is_newbie",
    "annual_from",
    "annual_to",
    "due_time",
    "skill_tags",
    "pay",
    "link",
    "deadline",
    "collected_at",
]

# 스키마를 명시한다: from_dicts 기본 추론이 첫 100행만 보는데,
# 소스별로 채워지는 컬럼이 달라(Wanted는 pay=NULL, work24는 문자열) 실데이터에서 타입 충돌이 난다.
SCHEMA = {
    "source": pl.String,
    "source_id": pl.String,
    "company": pl.String,
    "position": pl.String,
    "location": pl.String,
    "intro": pl.String,
    "main_tasks": pl.String,
    "requirements": pl.String,
    "preferred_points": pl.String,
    "benefits": pl.String,
    "employment_type": pl.String,
    "is_newbie": pl.Boolean,
    "annual_from": pl.Int64,
    "annual_to": pl.Int64,
    "due_time": pl.String,
    "skill_tags": pl.List(pl.String),
    "pay": pl.String,
    "link": pl.String,
    "deadline": pl.String,
    "collected_at": pl.Datetime("us", "UTC"),
}


def clean_text(value: Any) -> str | None:
    """HTML 태그 제거 + 공백 정규화. 비면 None."""
    if value is None:
        return None
    cleaned = re.sub(r"<[^>]+>", " ", str(value))
    cleaned = re.sub(r"\s+", " ", cleaned).strip()
    return cleaned or None


def normalize_wanted(record: dict[str, Any]) -> dict[str, Any] | None:
    payload = record["payload"]
    detail = payload.get("detail") or {}
    source_id = str(payload.get("id") or record.get("source_id") or "")
    if not source_id:
        return None
    return {
        "source": "wanted",
        "source_id": source_id,
        "company": clean_text((payload.get("company") or {}).get("name")),
        "position": clean_text(detail.get("position")),
        "location": clean_text((payload.get("address") or {}).get("full_location")),
        "intro": clean_text(detail.get("intro")),
        "main_tasks": clean_text(detail.get("main_tasks")),
        "requirements": clean_text(detail.get("requirements")),
        "preferred_points": clean_text(detail.get("preferred_points")),
        "benefits": clean_text(detail.get("benefits")),
        "employment_type": clean_text(payload.get("employment_type")),
        "is_newbie": payload.get("is_newbie"),
        "annual_from": payload.get("annual_from"),
        "annual_to": payload.get("annual_to"),
        "due_time": payload.get("due_time") or "상시채용",
        "skill_tags": [str(t) for t in (payload.get("skill_tags") or []) if t],
        "pay": None,
        "link": None,
        "deadline": None,
        "collected_at": record.get("collected_at"),
    }


def normalize_work24(record: dict[str, Any]) -> dict[str, Any] | None:
    payload = record["payload"]
    source_id = str(payload.get("source_id") or record.get("source_id") or "")
    if not source_id:
        return None
    return {
        "source": "work24",
        "source_id": source_id,
        "company": clean_text(payload.get("company")),
        "position": clean_text(payload.get("title")),
        "location": clean_text(payload.get("location")),
        "intro": None,
        "main_tasks": clean_text(payload.get("description")),
        "requirements": clean_text(payload.get("requirements")),
        "preferred_points": clean_text(payload.get("preferred")),
        "benefits": None,
        "employment_type": None,
        "is_newbie": None,
        "annual_from": None,
        "annual_to": None,
        "due_time": payload.get("deadline") or "채용시까지",
        "skill_tags": [],
        "pay": clean_text(payload.get("pay")),
        "link": payload.get("link"),
        "deadline": payload.get("deadline"),
        "collected_at": record.get("collected_at"),
    }


NORMALIZERS = {
    "wanted": normalize_wanted,
    "work24": normalize_work24,
}


def build_full_text(row: dict[str, Any]) -> str:
    """임베딩 입력용 단일 문서 텍스트 합성."""
    parts: list[str] = []
    if row.get("company"):
        parts.append(f"[회사] {row['company']}")
    if row.get("position"):
        parts.append(f"[포지션] {row['position']}")
    if row.get("annual_from") is not None or row.get("annual_to") is not None:
        parts.append(f"[경력요건] {row.get('annual_from') or 0}년 ~ {row.get('annual_to') or 0}년")
    if row.get("is_newbie"):
        parts.append("[신입 가능]")
    if row.get("skill_tags"):
        parts.append(f"[기술스택] {', '.join(row['skill_tags'])}")
    if row.get("main_tasks"):
        parts.append(f"[주요업무] {row['main_tasks']}")
    if row.get("requirements"):
        parts.append(f"[자격요건] {row['requirements']}")
    if row.get("preferred_points"):
        parts.append(f"[우대사항] {row['preferred_points']}")
    if row.get("location"):
        parts.append(f"[위치] {row['location']}")
    return "\n".join(parts)


def build_silver_frame(raw_records: list[dict[str, Any]]) -> pl.DataFrame:
    """bronze 레코드를 통합 스키마로 정제·중복제거한 DataFrame을 만든다."""
    normalized: list[dict[str, Any]] = []
    for record in raw_records:
        normalizer = NORMALIZERS.get(record["source"])
        if normalizer is None:
            logger.warning("알 수 없는 소스: %s", record["source"])
            continue
        try:
            row = normalizer(record)
        except Exception as error:
            logger.warning("정제 실패 (%s/%s): %s", record["source"], record["source_id"], error)
            continue
        # 회사/포지션이 모두 없는 레코드는 쓸모 없다
        if row is not None and (row["company"] or row["position"]):
            normalized.append(row)

    if not normalized:
        return pl.DataFrame(schema=dict.fromkeys(COLUMNS, pl.String))

    frame = pl.DataFrame(normalized, schema=SCHEMA)
    frame = frame.unique(subset=["source", "source_id"], keep="last", maintain_order=True)
    return frame.with_columns(
        pl.struct(COLUMNS)
        .map_elements(build_full_text, return_dtype=pl.String)
        .alias("full_text")
    )


UPSERT_SQL = """
INSERT INTO silver.jobs (
    source, source_id, company, position, location, intro, main_tasks,
    requirements, preferred_points, benefits, employment_type, is_newbie,
    annual_from, annual_to, due_time, skill_tags, pay, link, deadline,
    full_text, collected_at
)
VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21)
ON CONFLICT (source, source_id) DO UPDATE SET
    company = EXCLUDED.company,
    position = EXCLUDED.position,
    location = EXCLUDED.location,
    intro = EXCLUDED.intro,
    main_tasks = EXCLUDED.main_tasks,
    requirements = EXCLUDED.requirements,
    preferred_points = EXCLUDED.preferred_points,
    benefits = EXCLUDED.benefits,
    employment_type = EXCLUDED.employment_type,
    is_newbie = EXCLUDED.is_newbie,
    annual_from = EXCLUDED.annual_from,
    annual_to = EXCLUDED.annual_to,
    due_time = EXCLUDED.due_time,
    skill_tags = EXCLUDED.skill_tags,
    pay = EXCLUDED.pay,
    link = EXCLUDED.link,
    deadline = EXCLUDED.deadline,
    full_text = EXCLUDED.full_text,
    collected_at = EXCLUDED.collected_at,
    updated_at = now()
"""

INSERT_PARAMS = [
    "source", "source_id", "company", "position", "location",
    "intro", "main_tasks", "requirements", "preferred_points",
    "benefits", "employment_type", "is_newbie", "annual_from",
    "annual_to", "due_time", "skill_tags", "pay", "link",
    "deadline", "full_text", "collected_at",
]


async def run() -> int:
    raw_records = await fetch(
        "SELECT source, source_id, payload, collected_at FROM bronze.raw_postings"
    )
    if not raw_records:
        logger.warning("silver: bronze 데이터 없음")
        return 0

    frame = await asyncio.to_thread(build_silver_frame, [dict(r) for r in raw_records])
    if frame.is_empty():
        return 0

    rows = frame.to_dicts()
    await executemany(
        UPSERT_SQL,
        [tuple(row[col] for col in INSERT_PARAMS) for row in rows],
    )
    logger.info("silver upsert: %d건", len(rows))
    return len(rows)

File: internal/silver/silver.go
// Package silver는 bronze 원본 레코드를 통합 스키마로 정제하는 순수 함수들이다.
// 원본 pipeline/silver.py와 동일한 필드 매핑·텍스트 정리·full_text 합성을 수행한다.
package silver

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"regexp"
	"strconv"
	"strings"

	"github.com/Chae-JS/mentoai/internal/domain"
)

var (
	tagRe   = regexp.MustCompile(`<[^>]+>`)
	spaceRe = regexp.MustCompile(`\s+`)
)

// CleanText는 HTML 태그 제거 + 공백 정규화. 비면 nil.
func CleanText(value any) *string {
	if value == nil {
		return nil
	}
	text := spaceRe.ReplaceAllString(tagRe.ReplaceAllString(fmt.Sprint(value), " "), " ")
	trimmed := strings.TrimSpace(text)
	if trimmed == "" {
		return nil
	}
	return &trimmed
}

// payload 접근 헬퍼: JSON 디코딩 값(any)에서 타입별로 안전하게 꺼낸다.
func get(m map[string]any, keys ...string) any {
	var cur any = m
	for _, k := range keys {
		asMap, ok := cur.(map[string]any)
		if !ok {
			return nil
		}
		cur = asMap[k]
	}
	return cur
}

func str(m map[string]any, keys ...string) *string {
	return CleanText(get(m, keys...))
}

func rawStr(m map[string]any, keys ...string) string {
	if v := get(m, keys...); v != nil {
		s, _ := v.(string)
		return s
	}
	return ""
}

// idOf는 JSON 원시값(문자열·숫자)을 문자열 식별자로 바꾼다.
// Python의 str(payload.get(key))에 해당한다.
func idOf(m map[string]any, keys ...string) string {
	v := get(m, keys...)
	switch n := v.(type) {
	case string:
		return n
	case float64:
		if n == float64(int64(n)) {
			return strconv.FormatInt(int64(n), 10)
		}
		return fmt.Sprint(n)
	case nil:
		return ""
	default:
		return fmt.Sprint(n)
	}
}

func num(m map[string]any, keys ...string) *int {
	v := get(m, keys...)
	switch n := v.(type) {
	case float64:
		i := int(n)
		return &i
	case int:
		return &n
	case json.Number:
		if i, err := n.Int64(); err == nil {
			out := int(i)
			return &out
		}
	}
	return nil
}

func boolean(m map[string]any, keys ...string) *bool {
	if v, ok := get(m, keys...).(bool); ok {
		return &v
	}
	return nil
}

// NormalizeWanted는 Wanted 원본 payload를 통합 스키마로 매핑한다.
func NormalizeWanted(record domain.RawRecord) (*domain.SilverRow, error) {
	var payload map[string]any
	if err := json.Unmarshal(record.Payload, &payload); err != nil {
		return nil, err
	}
	sourceID := idOf(payload, "id")
	if sourceID == "" {
		sourceID = record.SourceID
	}
	if sourceID == "" {
		return nil, nil
	}
	dueTime := rawStr(payload, "due_time")
	if dueTime == "" {
		dueTime = "상시채용"
	}
	var skillTags []string
	if tags, ok := get(payload, "skill_tags").([]any); ok {
		for _, t := range tags {
			if t == nil {
				continue
			}
			s := fmt.Sprint(t)
			if s != "" {
				skillTags = append(skillTags, s)
			}
		}
	}
	return &domain.SilverRow{
		Source:         "wanted",
		SourceID:       sourceID,
		Company:        str(payload, "company", "name"),
		Position:       str(payload, "detail", "position"),
		Location:       str(payload, "address", "full_location"),
		Intro:          str(payload, "detail", "intro"),
		MainTasks:      str(payload, "detail", "main_tasks"),
		Requirements:   str(payload, "detail", "requirements"),
		PreferredPoint: str(payload, "detail", "preferred_points"),
		Benefits:       str(payload, "detail", "benefits"),
		EmploymentType: str(payload, "employment_type"),
		IsNewbie:       boolean(payload, "is_newbie"),
		AnnualFrom:     num(payload, "annual_from"),
		AnnualTo:       num(payload, "annual_to"),
		DueTime:        dueTime,
		SkillTags:      skillTags,
		Pay:            nil,
		Link:           nil,
		Deadline:       nil,
		CollectedAt:    record.CollectedAt,
	}, nil
}

// NormalizeWork24는 work24 원본 payload를 통합 스키마로 매핑한다.
func NormalizeWork24(record domain.RawRecord) (*domain.SilverRow, error) {
	var payload map[string]any
	if err := json.Unmarshal(record.Payload, &payload); err != nil {
		return nil, err
	}
	sourceID := idOf(payload, "source_id")
	if sourceID == "" {
		sourceID = record.SourceID
	}
	if sourceID == "" {
		return nil, nil
	}
	dueTime := rawStr(payload, "deadline")
	if dueTime == "" {
		dueTime = "채용시까지"
	}
	return &domain.SilverRow{
		Source:         "work24",
		SourceID:       sourceID,
		Company:        str(payload, "company"),
		Position:       str(payload, "title"),
		Location:       str(payload, "location"),
		Intro:          nil,
		MainTasks:      str(payload, "description"),
		Requirements:   str(payload, "requirements"),
		PreferredPoint: str(payload, "preferred"),
		Benefits:       nil,
		EmploymentType: nil,
		IsNewbie:       nil,
		AnnualFrom:     nil,
		AnnualTo:       nil,
		DueTime:        dueTime,
		SkillTags:      []string{},
		Pay:            str(payload, "pay"),
		Link:           str(payload, "link"),
		Deadline:       str(payload, "deadline"),
		CollectedAt:    record.CollectedAt,
	}, nil
}

var normalizers = map[string]func(domain.RawRecord) (*domain.SilverRow, error){
	"wanted": NormalizeWanted,
	"work24": NormalizeWork24,
}

// BuildFullText는 임베딩 입력용 단일 문서 텍스트를 합성한다.
func BuildFullText(row domain.SilverRow) string {
	var parts []string
	if row.Company != nil {
		parts = append(parts, fmt.Sprintf("[회사] %s", *row.Company))
	}
	if row.Position != nil {
		parts = append(parts, fmt.Sprintf("[포지션] %s", *row.Position))
	}
	if row.AnnualFrom != nil || row.AnnualTo != nil {
		from, to := 0, 0
		if row.AnnualFrom != nil {
			from = *row.AnnualFrom
		}
		if row.AnnualTo != nil {
			to = *row.AnnualTo
		}
		parts = append(parts, fmt.Sprintf("[경력요건] %d년 ~ %d년", from, to))
	}
	if row.IsNewbie != nil && *row.IsNewbie {
		parts = append(parts, "[신입 가능]")
	}
	if len(row.SkillTags) > 0 {
		parts = append(parts, fmt.Sprintf("[기술스택] %s", strings.Join(row.SkillTags, ", ")))
	}
	if row.MainTasks != nil {
		parts = append(parts, fmt.Sprintf("[주요업무] %s", *row.MainTasks))
	}
	if row.Requirements != nil {
		parts = append(parts, fmt.Sprintf("[자격요건] %s", *row.Requirements))
	}
	if row.PreferredPoint != nil {
		parts = append(parts, fmt.Sprintf("[우대사항] %s", *row.PreferredPoint))
	}
	if row.Location != nil {
		parts = append(parts, fmt.Sprintf("[위치] %s", *row.Location))
	}
	return strings.Join(parts, "\n")
}

// Build는 bronze 레코드 전체를 정제·중복제거(keep-last)한다.
// Polars unique(subset=[source, source_id], keep="last", maintain_order=True)와
// 동일하게, 각 키의 마지막 등장 위치에 마지막 내용을 남긴다.
func Build(records []domain.RawRecord) []domain.SilverRow {
	var normalized []domain.SilverRow
	for _, record := range records {
		normalize, ok := normalizers[record.Source]
		if !ok {
			slog.Warn("알 수 없는 소스", "source", record.Source)
			continue
		}
		row, err := normalize(record)
		if err != nil {
			slog.Warn("정제 실패", "source", record.Source, "source_id", record.SourceID, "error", err)
			continue
		}
		// 회사/포지션이 모두 없는 레코드는 쓸모 없다
		if row != nil && (isSet(row.Company) || isSet(row.Position)) {
			row.FullText = BuildFullText(*row)
			normalized = append(normalized, *row)
		}
	}

	// keep-last dedup: 뒤에서부터 first-seen만 남기고 뒤집는다.
	seen := make(map[string]struct{}, len(normalized))
	out := make([]domain.SilverRow, 0, len(normalized))
	for i := len(normalized) - 1; i >= 0; i-- {
		row := normalized[i]
		key := row.Source + "\x00" + row.SourceID
		if _, dup := seen[key]; dup {
			continue
		}
		seen[key] = struct{}{}
		out = append(out, row)
	}
	for i, j := 0, len(out)-1; i < j; i, j = i+1, j-1 {
		out[i], out[j] = out[j], out[i]
	}
	return out
}

func isSet(s *string) bool { return s != nil && *s != "" }
